### PR TITLE
Include `cmm-sources` when linking shared objects

### LINKED ISSUE #7182: `cmm-sources` are not included in shared object linking
**Describe the bug**
`cmm-sources` objects are not linked into shared objects.

**To Reproduce**

```
$ cd Cabal/cabal-testsuite/PackageTests/CmmSources
$ cabal v2-build --enable-executable-dynamic
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
In order, the following will be built (use -v for more details):
 - cmmexperiment-0 (exe:demo) (first run)
Preprocessing executable 'demo' for cmmexperiment-0..
Building executable 'demo' for cmmexperiment-0..
Linking /opt/exp/ghc/ghc-landing/libraries/Cabal/cabal-testsuite/PackageTests/CmmSources/dist-newstyle/build/x86_64-linux/ghc-8.8.3/cmmexperiment-0/x/demo/build/demo/demo ...
/opt/exp/ghc/ghc-landing/libraries/Cabal/cabal-testsuite/PackageTests/CmmSources/dist-newstyle/build/x86_64-linux/ghc-8.8.3/cmmexperiment-0/build/libHScmmexperiment-0-inplace-ghc8.8.3.so: error: undefined reference to 'aToMyWordzh'
collect2: error: ld returned 1 exit status
`cc' failed in phase `Linker'. (Exit code: 1)
```

Please use version-prefixed commands (e.g. `v2-build` or `v1-build`) to avoid ambiguity.

**Expected behavior**
A runnable executable should be producd.

**System information**
 - Any
 - `cabal` 5aea8a4b8463e1ae95272e190a1022764164294f

**Additional context**

### PR DESCRIPTION
**Pull request by bgamari**
---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog (add file to `changelog.d` directory).
* [x] The documentation has been updated, if necessary.

FIxes #7182 by fixing inconsistent object sets used during linking of vanilla and shared object output.

### CODE REVIEW

**Comment by phadej** [comment]
Should the test work on Windows? @Mistuke, are dynamic objects truly missing, or is there an issue with setup?

```
Perhaps you haven't installed the "dyn" libraries for package âbase-4.12.0.0â?
```

File: Cabal/Distribution/Simple/GHC.hs [hunk 1]
```diff
@@ -1364,7 +1364,9 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
                                              [tmpDir </> x | x <- cObjs ++ cxxObjs]
                     }
       dynLinkerOpts = mempty {
-                      ghcOptRPaths         = rpaths
+                      ghcOptRPaths         = rpaths,
+                      ghcOptInputFiles     = toNubListR
+                                             [tmpDir </> x | x <- cObjs ++ cxxObjs]
```
Comment (phadej) [inline Cabal/Distribution/Simple/GHC.hs:1369]: This change is something else? We don't need `cmmObjs` nor `asmObjs` however?

**Comment by Mistuke** [comment]
> Should the test work on Windows? @Mistuke, are dynamic objects truly missing, or is there an issue with setup?

No, sadly dynamic way is not supported on Windows. Table next to code makes it hard.. I still have it on my list of things to fix.

**Review by phadej — CHANGES_REQUESTED** [review]
> No, sadly dynamic way is not supported on Windows.

Then the new test should be skipped on windows. `skipIfWindows` (IIRC)

**Comment by phadej** [comment]
Please make PR against `master`, I'll backport them. I almost missed that again.

**Pull request closed without merging**